pkg/kv/transport.go

```diff
@@ -199,56 +207,47 @@ func (gt *grpcTransport) SendNext(ctx context.Context, done chan<- BatchCall) {
-				log.VEvent(ctx, 2, "sending request to local server")
-
-				// Clone the request. At the time of writing, Replica may mutate it
-				// during command execution which can lead to data races.
-				//
-				// TODO(tamird): we should clone all of client.args.Header, but the
-				// assertions in protoutil.Clone fire and there seems to be no
-				// reasonable workaround.
-				origTxn := client.args.Txn
-				if origTxn != nil {
-					clonedTxn := origTxn.Clone()
-					client.args.Txn = &clonedTxn
-				}
-
-				// Create a new context from the existing one with the "local request" field set.
-				// This tells the handler that this is an in-procress request, bypassing ctx.Peer checks.
-				localCtx := grpcutil.NewLocalRequestContext(ctx)
-
-				gt.opts.metrics.LocalSentCount.Inc(1)
-				return localServer.Batch(localCtx, &client.args)
-			}
-
-			log.VEventf(ctx, 2, "sending request to %s", client.remoteAddr)
-			reply, err := client.client.Batch(ctx, &client.args)
-			if reply != nil {
-				for i := range reply.Responses {
-					if err := reply.Responses[i].GetInner().Verify(client.args.Requests[i].GetInner()); err != nil {
-						log.Error(ctx, err)
-					}
-				}
-			}
-			return reply, err
-		}()
+		call := gt.send(ctx, client)
 		// NotLeaseHolderErrors can be retried.
 		var retryable bool
-		if reply != nil && reply.Error != nil {
+		if call.Reply != nil && call.Reply.Error != nil {
 			// TODO(spencer): pass the lease expiration when setting the state
 			// to set a more efficient deadline for retrying this replica.
-			if _, ok := reply.Error.GetDetail().(*roachpb.NotLeaseHolderError); ok {
+			if _, ok := call.Reply.Error.GetDetail().(*roachpb.NotLeaseHolderError); ok {
 				retryable = true
 			}
 		}
 		gt.setState(client.args.Replica, false /* pending */, retryable)
-		done <- BatchCall{Reply: reply, Err: err}
+		done <- call
 	}()
 }
 
+func (gt *grpcTransport) send(ctx context.Context, client batchClient) BatchCall {
+	gt.opts.metrics.SentCount.Inc(1)
+	if localServer := gt.rpcContext.GetLocalInternalServerForAddr(client.remoteAddr); localServer != nil {
+		log.VEvent(ctx, 2, "sending request to local server")
+
+		// Create a new context from the existing one with the "local request" field set.
+		// This tells the handler that this is an in-procress request, bypassing ctx.Peer checks.
+		localCtx := grpcutil.NewLocalRequestContext(ctx)
+
+		gt.opts.metrics.LocalSentCount.Inc(1)
+		reply, err := localServer.Batch(localCtx, &client.args)
+		return BatchCall{Reply: reply, Err: err}
+	}
+
+	log.VEventf(ctx, 2, "sending request to %s", client.remoteAddr)
+	reply, err := client.client.Batch(ctx, &client.args)
+	if reply != nil {
+		for i := range reply.Responses {
+			if err := reply.Responses[i].GetInner().Verify(client.args.Requests[i].GetInner()); err != nil {
+				log.Error(ctx, err)
+			}
+		}
+	}
+	return BatchCall{Reply: reply, Err: err}
```

Comment:
why not preserve the multiple return here? you lose tooling support for error checking without much of a benefit.